File: src/main/java/com/example/dpop/account/Account.java
package com.example.dpop.account;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "account")
public class Account {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long personId;

    private Instant createdAt;

    @JdbcTypeCode(SqlTypes.JSON)
    private List<AccountIdentification> identifications = new ArrayList<>();

    @JdbcTypeCode(SqlTypes.JSON)
    private List<AuthenticationMethod> authenticationMethods = new ArrayList<>();

    protected Account() {
    }

    public Account(Long personId, Instant createdAt) {
        this.personId = personId;
        this.createdAt = createdAt;
    }

    public Long getId() {
        return id;
    }

    public Long getPersonId() {
        return personId;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public List<AccountIdentification> getIdentifications() {
        return identifications;
    }

    public List<AuthenticationMethod> getAuthenticationMethods() {
        return authenticationMethods;
    }

    public void addIdentification(AccountIdentification identification) {
        identifications.add(identification);
    }

    public void addAuthenticationMethod(AuthenticationMethod authenticationMethod) {
        authenticationMethods.add(authenticationMethod);
    }
}

File: src/main/kotlin/com/example/dpop/account/internal/Account.kt
package com.example.dpop.account.internal

import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import jakarta.persistence.Version
import java.time.Instant

/**
 * Identity key and optimistic-lock root of an account - nothing else. Current state lives in rows
 * keyed by it ([AccountAnchor], [AccountAuthMethod]); history in append-only logs
 * ([AccountClaim], [AccountIdentification]).
 *
 * [version] serializes changes to the CURRENT state: `AccountService` loads the row with
 * `OPTIMISTIC_FORCE_INCREMENT` before touching anchors or methods, so two concurrent writers for
 * the same account cannot both succeed (the loser gets `409 CONCURRENT_MODIFICATION`). Log
 * appends never bump it.
 */
@Entity
@Table(schema = "account", name = "account")
class Account(
    var createdAt: Instant? = null
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null

    @Version
    var version: Long? = null
}
